### DISCUSSION TITLE
Определить место играющего трека в плейлисте

### DISCUSSION THREAD

**Discussion by Bedmit** [Chimildic/goofy#102] [Идея] [answered]
Вопрос состоит из двух.
Первый - возможно ли получить id активного плейлиста?
Второй - возможно ли определить позицию активного трека в этом плейлисте? Если первое не возможно, то в любом заданном.

Например реализуема ли задача, удаления из играющего плейлиста уже проигранных треков, т.е. тех которые находятся перед активным треком?
Решить ее просто методом удаления истории прослушивания, не очень получается. Часть треков в истории, в процессе вычислений меняются на похожие. Особенно это заметно сейчас, когда история идет только от Ласта, который частенько грешит подменой треков. В результате в очищаемом плейлисте остаются хвосты.

**Comment by Chimildic** [comment, marked as answer]
> от Ласта, который частенько грешит подменой треков

Даже если Last.fm сохранит оригинальное название, Spotify способен подменить тоже, ведь идет обратный поиск. Плюс никто баги не отменял. [ilyal](https://4pda.to/forum/index.php?showuser=675488) уже отловил парочку за эти дни)

Обновив [код библиотеки](https://script.google.com/d/1DnC4H7yjqPV2unMZ_nmB-1bDSJT9wQUJ7Wq-ijF4Nc7Fl3qnbT0FkPSr/edit?usp=sharing) и [права доступа](https://chimildic.github.io/goofy/#/install?id=Обновить-права-доступа), можно в обоих вопросах. Остается только на практике посмотреть мешается ли релинк в этом случае. От этого зависит как искать трек. И насколько стабильно вообще работает, учитывая недавний сбой получения истории. 

В `1.4.9` заменил функцию `getPlayingTrack` на [getPlayback](https://chimildic.github.io/goofy/#/func?id=getplayback). Предыдущий вариант давал лишь трек, новый принесет еще и контекст плеера. На случай если кто-то использовал `getPlayingTrack`, достать трек у `getPlayback` можно так. Если трека нет, объект пустой. 
```js
let playback = RecentTracks.getPlayback();
let track = playback.item;
```
Для парсинга _id_ понадобится [Helper](https://github.com/Chimildic/goofy/blob/main/addons/Запуск%20с%20телефона/helper.js), который есть у пользователей [управления с телефона](https://github.com/Chimildic/goofy/discussions/9).
```js
let playback = RecentTracks.getPlayback();
if (playback.isEmpty() || playback.context.type != 'playlist' || playback.shuffle_state || playback.repeat_state == 'track') {
  // прервать в ситуации когда нет данных, играет не плейлист, есть случайная перемешка или повтор трека
  return;
}

let playlistId = Helper.parseId(playback.context.uri);
let playlistTracks = Source.getPlaylistTracks('', playlistId);
let index = playlistTracks.findIndex(pt => pt.id == playback.item.id);
Selector.keepAllExceptFirst(playlistTracks, index > 0 ? index : 0);
```
Для удаления играющего трека тоже, достаточно прибавить единицу к индексу
```js
Selector.keepAllExceptFirst(playlistTracks, index + 1);
```

**Reply by Bedmit** [reply]
Отлично. Погонял немного, пока ни одной ошибки.
А помогите еще пожалуйста, как правильно вставить в проверку условие того, что играет конкретный плейлист, с заранее заданным ID. И условие, что активный трек изменился по сравнению с предыдущей проверкой.

**Reply by Bedmit** [reply]
И еще уточню. `index` это то число треков, которое мы удаляем. Если я хочу в конце плейлиста добавить такое же количество треков от плейлиста-донора, вместо числа добавляемых треков в аргументы просто написать `index`?

**Reply by Chimildic** [reply]
Понадобится [недавно](https://github.com/Chimildic/goofy/discussions/9#discussioncomment-980229) обновленный [Helper](https://github.com/Chimildic/goofy/blob/main/addons/Запуск%20с%20телефона/helper.js), где есть функции со `*State`. _id_ плейлиста в константу подставить.

```js
const ACTION_ID = 'ActionCleanPlaylist';
const PLAYLIST_ID = '';

let state = Helper.readState(ACTION_ID);
if (!state.hasOwnProperty('lastPlayingTrack')) {
  state.lastPlayingTrack = {};
}

let playback = RecentTracks.getPlayback();
if (playback.isEmpty()
  || playback.context.type != 'playlist'
  || !playback.context.uri.includes(PLAYLIST_ID)
  || playback.item.id == state.lastPlayingTrack.id
  || playback.repeat_state == 'track'
  || playback.shuffle_state) {
  return;
} else {
  let temp = [playback.item];
  Cache.compressTracks(temp);
  state.lastPlayingTrack = temp[0];
  Helper.updateState(ACTION_ID, state);
}

let playlistTracks = Source.getPlaylistTracks('', PLAYLIST_ID);
let index = playlistTracks.findIndex(pt => pt.id == playback.item.id);
Selector.keepAllExceptFirst(playlistTracks, index > 0 ? index : 0);
```

`index` это позиция играющего трека в массиве. Отсчет от нуля. Без перемешивания, `index` можно считать количеством треков не включая играющий. При `index + 1` включая.

Кстати, если трек не будет найден, `index` становится `-1`, что сломает алгоритм. Поэтому добавил однострочную проверку `index > 0 ? index : 0`.

**Reply by Bedmit** [reply]
Класс! Прямо как первый секс - ни чего не понял, но очень понравилось. Все работает именно так, как и хотел.

**Comment by Bedmit** [comment]
В связи с активным использованием разных устройств и переключением между ними, с помощью запуска Таскером скрипта типа
```js
function Auto() {
  const SMARTIFY_ID = '74uSA9FulwvWqVSi7mxWyQ';
  const DEVICE_ID = 'a051d9d7378f1ca083d9def8880bdd719cd37af0';

  let playback = Player.getPlayback();
  if (playback.isEmpty()) {
    Player.resume({ deviceId: DEVICE_ID, context_uri: `spotify:playlist:${SMARTIFY_ID}` });
  } else {
    Player.transferPlayback(DEVICE_ID, true);
  }
}
```
возникла необходимость в некоторой коррекции скрипта из этой темы.
Сейчас он выглядит так 
```js
function Smartify() {
  const ACTION_ID = 'ActionCleanPlaylist';
  const PLAYLIST_ID = '74uSA9FulwvWqVSi7mxWyQ';

  let state = Helper.readState(ACTION_ID);
  if (!state.hasOwnProperty('lastPlayingTrack')) {
    state.lastPlayingTrack = {};
  }

  let playback = Player.getPlayback();
  if (playback.isEmpty()
    || !playback.context.uri.includes(PLAYLIST_ID)
    || playback.item.id == state.lastPlayingTrack.id
    || playback.repeat_state == 'track'
    || playback.shuffle_state) {
    return;
  } else {
    let temp = [playback.item];
    Cache.compressTracks(temp);
    state.lastPlayingTrack = temp[0];
    Helper.updateState(ACTION_ID, state);
  }

  let playlistTracks = Source.getPlaylistTracks('', PLAYLIST_ID);
  let index = playlistTracks.findIndex(pt => pt.id == playback.item.id);

  Selector.keepAllExceptFirst(playlistTracks, index > 0 ? index : 0); // треки с текущим и далее

  let Smartify = Cache.read('Smartify.json'); // получение плейлиста-донора
  Combiner.push(playlistTracks, Smartify); // остаток + треки из плейлиста-донора
  Selector.keepFirst(playlistTracks, 30);
  Filter.removeTracks(Smartify, playlistTracks);
  Cache.write('Smartify.json', Smartify); // запись в кеш остатка плейлиста-донора

  let Juice = Cache.read('Juice.json'); // получение и добавление в кэш проигранных треков плейлиста лучших треков
  Filter.removeTracks(Juice, playlistTracks, true);
  Cache.append('OldJuice.json', Juice, 'begin');

  let Minions = Cache.read('Minions.json'); // получение и добавление в кэш проигранных избранных треков любимых артистов 
  Filter.removeTracks(Minions, playlistTracks, true);
  Cache.append('OldMinions.json', Minions, 'begin');

  let UpdTime = 'Обновлено в ' + new Date().toTimeString().substring(0, 5);
  Playlist.saveWithReplace({ id: '74uSA9FulwvWqVSi7mxWyQ', name: 'Smartify', description: UpdTime, public: false, tracks: playlistTracks });
}
```
Скрипт постоянно отрезает из плейлиста Smartify уже проигранные треки. Это нужно в основном для того, чтобы в случае переключения/включения другого устройства, при отсутствии плэйбэка запускать плейлист Smartify, с трека игравшего последним. 

Но по условию скрипт не всегда обрезает плейлист. Например в момент срабатывания триггера играл трек не из плейлиста, а из очереди воспроизведения. Или вообще плейлист стоит на паузе и/или отсутствует плэйбек.
В результате два-три уже проигранных трека не отрезались и при запуске Smartify откатывается назад и начинает воспроизведение с них, а не с того, что играл последним.

Можно ли как-то модернизировать скрипт, чтобы он отрезал проигранные треки до последнего запомненного трека (я так понимаю он пишется в кэш?) в любом случае?

**Reply by Chimildic** [reply]
Уже подзабыл зачем понадобился `state`. Что если смотреть плейбэк и при нужном плейлисте сразу удалять? Все равно кроме него ничто не скажет что удалять.
```js
  const PLAYLIST_ID = '74uSA9FulwvWqVSi7mxWyQ';

  let playback = Player.getPlayback();
  let isTargetPls = playback.context && playback.context.uri.includes(PLAYLIST_ID);
  if (!isTargetPls
    || (isTargetPls && playback.repeat_state == 'track')
    || (isTargetPls && playback.shuffle_state)) {
    return;
  }

  let playlistTracks = Source.getPlaylistTracks('', PLAYLIST_ID);
  let index = playlistTracks.findIndex(pt => pt.id == playback.item.id);
  if (index == -1) {
    console.log('Трек не найден или уже удален');
    return;
  }
  Selector.keepAllExceptFirst(playlistTracks, index);

  Playlist.saveWithReplace({
    id: '74uSA9FulwvWqVSi7mxWyQ',
    name: 'Smartify',
    description: 'Обновлено в ' + new Date().toTimeString().substring(0, 5),
    public: false,
    tracks: playlistTracks
  });
```

**Reply by Bedmit** [reply]
> подзабыл зачем понадобился `state`

Ах-ха, а я и не знал ))
Подозреваю для проверки того, что плейбэк изменился относительно прошлого. Чтобы не гонять скрипт вхолостую на паузе. 


> смотреть плейбэк и при нужном плейлисте сразу удалять

Я бы плейбэк и на принадлежность к плейлисту Smartify не проверял. Чтобы проигранные треки удалялись и во время играющих из очереди воспроизведения треков. Но как их тогда вычислять?
Разве что использовать для ведения последней кратковременной истории плейбэк, раз в минуту добавляемый в кэш. Реже  не имеет смысла. И просто удалять последние треки этого кэша из плейлиста. Не знаю на сколько это будет затратно. 

Пока просто уменьшил триггер скрипта до пяти минут. Посмотрю, не упрусь ли в лимит запросов.

**Reply by Chimildic** [reply]
Плебэк говорит что играет здесь и сейчас. Кэш тут не нужен. Когда сработает триггер, целевой трек будет найден.
Если не проверить принадлежность, случайный трек совпадёт с плейлистом и удалится все, что на самом деле не было прослушано.
Единственная альтернатива это удаление по истории прослушиваний.

**Reply by Bedmit** [reply]
> Плебэк говорит что играет здесь и сейчас

Это я знаю. А дальше идёт проверка его наличия в плейлисте. Но проблема в том, что у меня частенько при прослушивании этого плейлиста играют треки не из него, а из очереди. Соответственно плейбэк определяется как не принадлежащий плейлисту и скрипт не срабатывает. 
Таким образом у меня может не удалиться несколько уже проигранных треков. И при большой паузе, когда плейбэка не будет, плейлист начнет играть сначала, опять проигрывая эти треки.
Удалять проигранные треки используя историю тоже не вариант. Там часто заменяются треки, да и 15 минут обновления слишком редко для использования в скрипте срабатывающем раз в 10 минут. 
Поэтому и предполагал как-то использовать вместо неё плейбэки записанные в кэш. 

Но все это теперь уже не важно. Я нашел решение проблемы с другой стороны. 
Т.к. в основном треки появляются в очереди при условиях, которые я инициирую с помощью Таскера, то к выполняемым задачам условий, я просто добавил задачу запуска этого скрипта.
